### PR TITLE
KOGITO-402: KOGITO-403: [DMN Designer] Close docks by default

### PR DESCRIPTION
**Pull request by manstis**
See:-

- https://issues.jboss.org/browse/KOGITO-402
- https://issues.jboss.org/browse/KOGITO-403

One PR, two commits (not to be squashed) one for each JIRA (403 relied on 402 so I've combined).

This replaces:
- https://github.com/kiegroup/kie-wb-common/pull/2936
- https://github.com/kiegroup/kie-wb-common/pull/2937

### CODE REVIEW

**Comment by manstis** [comment]
@jomarko To test you'll need to build `kie-wb-common` with `-Pkogito` (until #2968 is merged).

Then run `mvn clean process-resources gwt:run` on `kie-wb-common-webapp-kogito-testing`.

**Comment by manstis** [comment]
@jomarko This PR is good to review.

**Comment by jomarko** [comment]
@manstis not able to build neither **stunner** nor **dmn** submodules, getting errors like:
```
[ERROR] Line 61: No source code is available for type org.uberfire.ext.widgets.core.client.editors.texteditor.TextEditorView; did you forget to inherit a required module?
[INFO]    [ERROR] Aborting compile due to errors in some input files
```

**Comment by jomarko** [comment]
Stopping code review as the code needs to be changed.

**Comment by manstis** [comment]
@jomarko This is ready for re-review.. Reported issue resolved.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDock.java [hunk 1]
```diff
@@ -59,11 +60,20 @@ public DecisionNavigatorDock(final UberfireDocks uberfireDocks,
...
     }
 
+    @Override
     public void init(final String perspective) {
         this.perspective = perspective;
         this.uberfireDock = makeUberfireDock();
     }
 
+    /**
+     * No operation. Destruction is handled by the close() method.
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDock.java]: Understand, however shouldn't we split responsibilities now when we have introduced `destroy` method?

**Comment by jomarko** [comment]
### Manual review
I do not know if can approve, VSCode extension still doesn't work for me, in testing app I am able to reproduce the issue.

Screenshot from 2019-10-30 14-11-11

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDock.java (hunk 1 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDock.java]: Fair point.. I've split the concerns in a new commit.

**Comment by manstis** [comment]
@jomarko It looks like you did not get the latest code, as I see the duplicate icons on the dock in your screen shot (also, when I run the `-testing` webapp I don't have any docks on the screen that lists diagrams; the dock only appears when I open the editor).

**_No dock (on right hand side)_**
Screenshot from 2019-10-30 16-37-42

**_Dock (on right hand side)_**
Screenshot from 2019-10-30 16-37-55

**Comment by jomarko** [comment]
@manstis may I ask you to resolve conflicts?

**Comment by manstis** [comment]
@jomarko Conflict resolved.

I looked at Sonar's complaint of duplicate code and unfortunately I cannot move to a shared super class as each of the _kogito_ docks extends a different base class already.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDockTest.java [hunk 2]
```diff
@@ -164,6 +164,18 @@ public void testCloseWhenItIsNotOpened() {
         verify(uberfireDocks, never()).remove(uberfireDock);
     }
 
+    @Test
+    public void testDestroy() {
+        final UberfireDock uberfireDock = mock(UberfireDock.class);
+
+        dock.setOpened(false);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDockTest.java]: I do not understand why we need this method, could you please clarify?

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoDecisionNavigatorDock.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public void close() {
+        if (!isOpened()) {
+            return;
+        }
+
+        isOpened = false;
+        uberfireDocks.close(getUberfireDock());
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoDecisionNavigatorDock.java]: Do not need we call `destroy()` here?

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPreviewDiagramDock.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public void close() {
+        if (!isOpened()) {
+            return;
+        }
+
+        isOpened = false;
+        uberfireDocks.close(getUberfireDock());
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPreviewDiagramDock.java]: Do not need we call `destroy()` here?

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPropertiesDock.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    public void close() {
+        if (!isOpened()) {
+            return;
+        }
+
+        isOpened = false;
+        uberfireDocks.close(getUberfireDock());
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPropertiesDock.java]: Do not need we call `destroy()` here?

**Comment by jomarko** [comment]
@manstis here is what I do in `webapp-kogito-testing` https://issues.jboss.org/secure/attachment/12457096/twodmns.webm

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDockTest.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDockTest.java]: It is not needed; removed. The copy and paste demons got me!

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoDecisionNavigatorDock.java (hunk 3 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoDecisionNavigatorDock.java]: In _kogito_ I tie the life-cycle of the dock to that of the editor so I don't destroy it when it is simply closed (but still visible as an icon on the dock sidebar). In Business Central the life-cycle of a dock is tied to it being opened and closed which leads to problems if we want the dock sidebar to have the icon but the dock itself to be closed.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoDecisionNavigatorDock.java]: So, no, I don't need to call `destroy()` here as the editor calls `destroy()` on the dock when the editor is closed.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPreviewDiagramDock.java (hunk 4 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPreviewDiagramDock.java]: As above.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPropertiesDock.java (hunk 5 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/docks/KogitoPropertiesDock.java]: As above.

**Comment by manstis** [comment]
@jomarko This is ready for re-review when you get a moment. I've fixed the issue.

**Comment by manstis** [comment]
@jomarko I am sorry for wasting your valuable time today.

I tried to be smart, moving dock handling to Focus/LostFocus events but when the content of the dock is focused it obviously (from your video) led to the Editor loosing focus and killing the dock content. I've restored behaviour to the Editor's open/close lifecycle states...

It remains flawed when opening/creating multiple DMN files and using the drop-down screen selector on the top left of the screen.. However the `-webapp-kogito-testing` webapp is really only to help me (and you!) fiddle around with how the DMN editor would be in kogito w/o having to build the VSCode plugin. In reality the workflow is much more simple: create a file, edit it, close it or open a file, edit it, close it. Both on a single file at a single time. I think `-webapp-kogito-testing` is good enough for this. 

This PR was about having the docks on the right hand side and closed by default. I hope this works OK for you and, ignoring the flawed operation when trying to open/edit multiple files at once.

**Comment by manstis** [comment]
@jomarko I consider this PR ready for review. See my final comments [here](https://github.com/kiegroup/kie-wb-common/pull/2971#issuecomment-548956310).

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by jomarko** [comment]
If I have more DMN files, and switching between them without closing, still able to reach state shown in the picture, however I am not going to invest more re-reviews into this PR.

Screenshot from 2019-11-06 11-33-28

**Pull request merged**